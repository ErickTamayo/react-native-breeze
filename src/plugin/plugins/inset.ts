import { PluginFunction } from "./types";

export type PluginGroups = {
  negative?: "-";
  position:
    | "inset"
    | "inset-x"
    | "inset-y"
    | "top"
    | "right"
    | "bottom"
    | "left"
    | "start"
    | "end";
  value: string;
};

export const pattern = /^(?<negative>-)?(?<position>(inset(-[xy])?)|top|right|bottom|left|start|end)?-?(?<value>\d+)$/;

export const plugin: PluginFunction<PluginGroups> = ({
  input,
  groups,
  theme,
}) => {
  const { negative, position, value } = groups;

  const key = `${negative || ""}${value}`;
  const number = theme(["inset", key]);

  if (typeof number !== "number") {
    console.error(`Invalid value [${typeof number}] for ${input}`);
    return {};
  }

  if (position === "inset") {
    return { top: number, right: number, bottom: number, left: number };
  }

  if (position === "inset-x") {
    return { right: number, left: number };
  }

  if (position === "inset-y") {
    return { top: number, bottom: number };
  }

  if (position === "top") {
    return { top: number };
  }

  if (position === "right") {
    return { right: number };
  }

  if (position === "bottom") {
    return { bottom: number };
  }

  if (position === "left") {
    return { left: number };
  }

  if (position === "start") {
    return { start: number };
  }

  if (position === "end") {
    return { end: number };
  }

  return {};
};
