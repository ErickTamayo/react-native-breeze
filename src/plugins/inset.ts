import { withNegativePrefix } from "../helpers/misc";
import { PluginFunction, PluginPattern } from "./types";

// prettier-ignore
export type PluginGroups = {
  prefix?: "-";
  position: "inset" | "inset-x" | "inset-y" | "top" | "right" | "bottom" | "left" | "start" | "end";
  key: string;
};

export const pattern: PluginPattern = ({ keys }) => {
  // prettier-ignore
  const positions = ["inset", "inset-x", "inset-y", "top", "right", "bottom", "left", "start", "end"].join('|');
  // prettier-ignore
  return new RegExp(`^(?<prefix>-)?(?<position>${positions})-(?<key>${keys("inset")})$`);
};

export const plugin: PluginFunction<PluginGroups> = ({
  input,
  groups,
  theme,
}) => {
  const { prefix, position, key } = groups;
  const value = theme(["inset", withNegativePrefix(prefix, key)]);

  switch (position) {
    case "inset-x":
      return { right: value, left: value };
    case "inset-y":
      return { top: value, bottom: value };
    case "top":
      return { top: value };
    case "bottom":
      return { bottom: value };
    case "left":
      return { left: value };
    case "right":
      return { right: value };
    case "start":
      return { start: value };
    case "end":
      return { end: value };
    default:
      return { top: value, right: value, bottom: value, left: value };
  }
};
