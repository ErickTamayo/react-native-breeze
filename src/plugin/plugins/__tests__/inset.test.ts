import {
  shouldEvaluateTheCorrectPatternTest,
  shouldMatchOutputSnapshot,
} from "../../utils/tests";
import { pattern, plugin } from "../inset";

describe("inset", () => {
  // prettier-ignore
  const shouldMatch = [ "inset-0", "inset-y-0", "inset-x-0", "top-0", "right-0", "left-0", "bottom-0", "start-0", "end-0"];

  shouldEvaluateTheCorrectPatternTest(pattern, {
    shouldMatch,
    shouldNotMatch: [
      "inset",
      "inset-y",
      "inset-x",
      "left",
      "right",
      "bottom",
      "start",
      "end",
    ],
  });

  shouldMatchOutputSnapshot(pattern, plugin, shouldMatch);
});
