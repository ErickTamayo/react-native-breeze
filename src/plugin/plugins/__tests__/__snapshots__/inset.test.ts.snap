// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inset Should parse (bottom-0) correctly 1`] = `
Object {
  "bottom": 0,
}
`;

exports[`inset Should parse (end-0) correctly 1`] = `
Object {
  "end": 0,
}
`;

exports[`inset Should parse (inset-0) correctly 1`] = `
Object {
  "bottom": 0,
  "left": 0,
  "right": 0,
  "top": 0,
}
`;

exports[`inset Should parse (inset-x-0) correctly 1`] = `
Object {
  "left": 0,
  "right": 0,
}
`;

exports[`inset Should parse (inset-y-0) correctly 1`] = `
Object {
  "bottom": 0,
  "top": 0,
}
`;

exports[`inset Should parse (left-0) correctly 1`] = `
Object {
  "left": 0,
}
`;

exports[`inset Should parse (right-0) correctly 1`] = `
Object {
  "right": 0,
}
`;

exports[`inset Should parse (start-0) correctly 1`] = `
Object {
  "start": 0,
}
`;

exports[`inset Should parse (top-0) correctly 1`] = `
Object {
  "top": 0,
}
`;
